### PR TITLE
Add reflect, symmetric and edge padding

### PR DESCRIPTION
**Pull request by utkuozbulak**
Added reflect, symmetric and edge padding options.

This option was requested by #400. PR #213 was sent to include the same functionality but still awaiting response from the author since Sep 20, 2017.

The new padding options also work the same way with how it is for the constant one.
If `padding` is an int: it pads all sides.
if `padding` is a tuple with two elements: it pads left/right and top/bottom respectively.
if `padding` is a tuple with four elements: it pads left, top, right, bottom borders respectively.
It also does not break or change the behavior of existing code.

In #213  @alykhantejani suggested that the function should be
`pad(img, padding,  padding_mode='constant', fill=0)` instead of 
`pad(img, padding, fill=0, padding_mode='constant')` because when the `padding_mode` is not _constant_, fill is irrelevant/not used. This is correct, but doing so might break existing code for some people so I did not change the positions.

Below are some tests:
RGB test image:
test_image

`pad(img, 20, padding_mode = 'reflect')`
20reflect

`pad(img, (0, 50), padding_mode = 'edge')`
0-50-edge

`pad(img, (1, 10, 20, 30), padding_mode = 'symmetric')`

Grayscale test image:
cat

`pad(img, (5,10,20,50), padding_mode = 'symmetric')`

5-10-20-50-sym

### CODE REVIEW

File: torchvision/transforms/transforms.py [hunk 1]
```diff
@@ -231,15 +231,17 @@ class Pad(object):
             length 3, it is used to fill R, G, B channels respectively.
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: Can you add docs here as well?

File: torchvision/transforms/functional.py [hunk 2]
```diff
@@ -220,9 +220,10 @@ def pad(img, padding, fill=0):
             on left/right and top/bottom respectively. If a tuple of length 4 is provided
             this is the padding for the left, top, right and bottom borders
             respectively.
-        fill: Pixel fill value. Default is 0. If a tuple of
+        fill: Pixel fill value for constant fill. Default is 0. If a tuple of
             length 3, it is used to fill R, G, B channels respectively.
-
+        padding_mode: Type of padding. Should be: constant, edge, reflect or symmetric.
```
Comment (fmassa) [inline torchvision/transforms/functional.py]: Can you expand on the differences between `reflect` and `symmetric`? From looking at the images, it was not clear to me what were the differences.

**Comment by utkuozbulak** [comment]
Inspired from what you suggested as the test case Included the test as well. It basically does following:

1- Create an empty tensor (image) with zeros
image

2- Change leftmost column to one.
image

3- Pad image with a constant value with a length of 1
image

4- Pad image with edge, symmetric or reflect (3 different cases, below only illustrated _symmetric_ version)
image

Then check the values of leftmost middle crop with the length of six if it is indeed symmetric, highlighted in the box above. Then check the shape of the matrix if it adds up to what we are expecting (Here, I also test if the padding is a tuple instead of a single value in different versions of padding).

File: test/test_transforms.py [hunk 3]
```diff
@@ -235,6 +235,37 @@ def test_pad_with_tuple_of_pad_values(self):
...
+        img[:, :, 0] = 1  # Constant value added to leftmost edge
+        img = transforms.ToPILImage()(img)
+        img = F.pad(img, 1, (200, 200, 200))
+
+        # pad 3 to all sidess
+        edge_padded_img = F.pad(img, 3, padding_mode='edge')
+        # First 6 elements of leftmost edge in the middle of the image, values are in order:
+        # edge_pad, edge_pad, edge_pad, constant_pad, constant value added to leftmost edge, 0
+        edge_middle_slice = np.asarray(edge_padded_img).transpose(2, 0, 1)[0][17][:6]
+        assert np.all(edge_middle_slice == np.asarray([200, 200, 200, 200, 255, 0]))
```
Comment (vfdev-5) [inline test/test_transforms.py]: This check is correct but somehow weird due to 
```python
img = torch.zeros(3, 27, 27)  # Float32
img[:, :, 0] = 1   # we add 1 and not 255
img = transforms.ToPILImage()(img)  # This converts 1 to 255 due to [pic = pic.mul(255).byte()](https://github.com/pytorch/vision/blob/master/torchvision/transforms/functional.py#L107)
```
maybe better to use 
```python
img = torch.zeros(3, 27, 27).byte()
```
and test with correct values [..., 200, 1, 0]. Same below

Comment (fmassa) [reply test/test_transforms.py]: If it can be clearer, can you send a PR?
